File: core/src/nl/t64/game/rpg/components/loot/Loot.java
package nl.t64.game.rpg.components.loot;

import lombok.Getter;

import java.util.Map;


@Getter
public class Loot {

    private static final String BONUS_PREFIX = "bonus_";

    private Map<String, Integer> content;
    private int trapLevel;
    private int lockLevel;
    private int xp;

    public Loot(Map<String, Integer> content) {
        this.content = content;
        this.trapLevel = 0;
        this.lockLevel = 0;
        this.xp = 0;
    }

    Loot() {
        this.content = Map.of();
        this.trapLevel = 0;
        this.lockLevel = 0;
        this.xp = 0;
    }

    public boolean isTaken() {
        return content.isEmpty();
    }

    public void clearContent() {
        content = Map.of();
    }

    public void updateContent(Map<String, Integer> newContent) {
        content = newContent;
    }

    public boolean isTrapped() {
        return trapLevel > 0;
    }

    public boolean canDisarmTrap(int mechanicLevel) {
        return mechanicLevel >= trapLevel;
    }

    public void disarmTrap() {
        trapLevel = 0;
    }

    public boolean isLocked() {
        return lockLevel > 0;
    }

    public boolean canPickLock(int thiefLevel) {
        return thiefLevel >= lockLevel;
    }

    public void pickLock() {
        lockLevel = 0;
    }

    public void clearXp() {
        xp = 0;
    }

    public boolean isXpGained() {
        return xp == 0;
    }

    public void handleBonus() {
        content.entrySet()
               .stream()
               .filter(entry -> entry.getKey().startsWith(BONUS_PREFIX))
               .forEach(bonusEntry -> handleBonus(bonusEntry.getKey(), bonusEntry.getValue()));
        removeBonus();
    }

    public void removeBonus() {
        content.keySet()
               .removeIf(itemId -> itemId.startsWith(BONUS_PREFIX));
    }

    private void handleBonus(String bonusItemId, int bonusAmount) {
        String itemId = bonusItemId.substring(BONUS_PREFIX.length());
        if (content.containsKey(itemId)) {
            int amount = content.get(itemId);
            content.put(itemId, amount + bonusAmount);
        } else {
            content.put(itemId, bonusAmount);
        }
    }

}

File: core/src/nl/t64/game/rpg/components/loot/Loot.kt
package nl.t64.game.rpg.components.loot


private const val BONUS_PREFIX = "bonus_"

class Loot(
    var content: MutableMap<String, Int> = mutableMapOf(),
    var trapLevel: Int = 0,
    var lockLevel: Int = 0,
    var xp: Int = 0
) {

    fun isTaken(): Boolean =
        content.isEmpty()

    fun clearContent() {
        content = mutableMapOf()
    }

    fun updateContent(newContent: MutableMap<String, Int>) {
        content = newContent
    }

    fun isTrapped(): Boolean =
        trapLevel > 0

    fun canDisarmTrap(mechanicLevel: Int): Boolean {
        return mechanicLevel >= trapLevel
    }

    fun disarmTrap() {
        trapLevel = 0
    }

    fun isLocked(): Boolean =
        lockLevel > 0

    fun canPickLock(thiefLevel: Int): Boolean {
        return thiefLevel >= lockLevel
    }

    fun pickLock() {
        lockLevel = 0
    }

    fun clearXp() {
        xp = 0
    }

    fun isXpGained(): Boolean =
        xp == 0

    fun handleBonus() {
        content.entries
            .filter { it.key.startsWith(BONUS_PREFIX) }
            .forEach { handleBonus(it.key, it.value) }
        removeBonus()
    }

    fun removeBonus() {
        content.keys.removeIf { it.startsWith(BONUS_PREFIX) }
    }

    private fun handleBonus(bonusItemId: String, bonusAmount: Int) {
        val itemId = bonusItemId.substring(BONUS_PREFIX.length)
        if (content.containsKey(itemId)) {
            val amount = content[itemId]!!
            content[itemId] = amount + bonusAmount
        } else {
            content[itemId] = bonusAmount
        }
    }

}
